File: OrangesRotting.java
// Time Complexity : O(n+n*m) for traversing all the elements to find freshoranges and a rotten orange
// Space Complexity : O(n*m) for queue
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// First traverse through the matrix for finding the number of fresh oranges and rotten oranges
// add all rotten oranges indexes to the queue.
// while the queus is not emplty pop the first index and find it's surrounding cells in 1 distances to find fresh orange
// Increment the counter for minutes once it enters the while loop of queue.
// for this use direction matrix which is intitialized below and change it to 2 (rotten) if present.
// add those indexes also to the queue to find further rotting oranges because of them.
// if there are no rotting oranges or no fresh oranges we will return -1


class Solution {
    public int orangesRotting(int[][] grid) {
        int freshOranges = 0;
        int min = 0;
        
        Queue<int[]> q = new LinkedList<>();
        
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(grid[i][j]==1){
                    freshOranges++;
                }
                else if(grid[i][j]==2){
                    q.add(new int[]{i,j});
                }
            }
        }
       
        
        int[][] dir = {{1,0},{-1,0},{0,1},{0,-1}};
        while(!q.isEmpty()){ 
            if(freshOranges==0) {return min;}
            min++; 
            int size = q.size();
            for(int k=0; k<size;k++){
                int[] pop = q.poll();
                for(int[] arr:dir){
                    int r=arr[0];
                    int c=arr[1];
                    int i=pop[0];
                    int j = pop[1];
                    if(i+r<grid.length && i+r>=0 && j+c<grid[0].length && j+c>=0 && grid[i+r][j+c]==1 ){
                    grid[i+r][j+c] = 2;
                    q.add(new int[]{i+r,j+c});
                    freshOranges--;
                    }
                }
                
            }
        }
        if(freshOranges>0) {return -1;}
        return min;
    }
}

File: orangesRotting.py
#Time Complexity:O(n), n being the total number of elements in the matrix
#Space Complexity: O(n)
#Ran successfully on Leetcode:Yes

#BFS Solution:
#Algorithm:
# 1. Calculate the length of rows and cols of matrix. 
# 2. Iterate through it and count fresh oranges, and if they are rotten , add then m to queue created. 
# 3. Set time ==0
# 4. Now pop pout of queue and store in curr variable, check its neighbour if there are fresh oranges, and convert them to rotten by assigning 2 and add their location to queue.
# 5. At the end if the fresh array is zero, we return time, else -1



class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        fresh=0
        if not grid or len(grid)==0:
            return fresh
        q=[]
        m=len(grid)
        n=len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    fresh+=1
                if grid[i][j]==2:
                    q.append((i,j))
        print(fresh)
        time=0
        if fresh==0:
             return 0
        while(q):
            size=len(q)
            for _ in range(size):
                curr=q.pop(0)
                for i,j in [(1,0),(0,1),(-1,0),(0,-1)]:
                    r=curr[0]+i
                    c=curr[1]+j
                    if  0<=r<m and 0<=c<n and grid[r][c]==1:
                        grid[r][c]=2
                        q.append((r,c))
                        fresh-=1
            time+=1
        if fresh==0:
            return time-1
        else:
            return -1
                

        
